resolver/addressable_resolver.go

```diff
@@ -86,128 +86,206 @@ func (r *URIResolver) URIFromDestination(ctx context.Context, dest duckv1beta1.D
+	obj, err := lister.ByNamespace(dest.Ref.Namespace).Get(dest.Ref.Name)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get object %s/%s: %w", ref.Namespace, ref.Name, err)
+		return nil, fmt.Errorf("failed to get object %s/%s: %w", dest.Ref.Namespace, dest.Ref.Name, err)
 	}
 
 	// K8s Services are special cased. They can be called, even though they do not satisfy the
 	// Callable interface.
-	if ref.APIVersion == "v1" && ref.Kind == "Service" {
+	if dest.Ref.APIVersion == "v1" && dest.Ref.Kind == "Service" {
 		url := &apis.URL{
 			Scheme: "http",
-			Host:   network.GetServiceHostname(ref.Name, ref.Namespace),
+			Host:   network.GetServiceHostname(dest.Ref.Name, dest.Ref.Namespace),
 			Path:   "",
 		}
-		return url, nil
+		if dest.CACerts != nil && *dest.CACerts != "" {
+			url.Scheme = "https"
+		}
+		return &duckv1.Addressable{
+			Name:    dest.Ref.Address,
+			URL:     url,
+			CACerts: dest.CACerts,
+		}, nil
 	}
 
 	addressable, ok := obj.(*duckv1.AddressableType)
 	if !ok {
-		return nil, apierrs.NewBadRequest(fmt.Sprintf("%+v (%T) is not an AddressableType", ref, ref))
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("%+v (%T) is not an AddressableType", dest.Ref, dest.Ref))
 	}
-	if addressable.Status.Address == nil {
-		return nil, apierrs.NewBadRequest(fmt.Sprintf("address not set for %+v", ref))
+
+	addr, err := r.selectAddress(dest, addressable)
+	if err != nil {
+		return nil, err
+	}
+	// Do not modify informer copy.
+	addr = addr.DeepCopy()
+
+	if addr.URL == nil {
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("URL missing in address of %+v", dest.Ref))
+	}
+	if addr.URL.Host == "" {
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("hostname missing in address of %+v", dest.Ref))
+	}
+
+	if addr.CACerts == nil {
+		addr.CACerts = dest.CACerts
+	}
+
+	return addr, nil
+}
+
+// selectAddress selects a single address from the given AddressableType
+func (r *URIResolver) selectAddress(dest duckv1.Destination, addressable *duckv1.AddressableType) (*duckv1.Addressable, error) {
+	if len(addressable.Status.Addresses) == 0 && addressable.Status.Address == nil {
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("address not set for %+v", dest.Ref))
 	}
-	url := addressable.Status.Address.URL
-	if url == nil {
-		return nil, apierrs.NewBadRequest(fmt.Sprintf("URL missing in address of %+v", ref))
+
+	// if dest.ref.address is specified:
+	// - select the first (in order) address from status.addresses with name == dest.ref.address
+	// - If no address is found return an error explaining that the address with name dest.ref.address
+	//   cannot be found/resolved
+	if dest.Ref.Address != nil && *dest.Ref.Address != "" {
+		for _, addr := range addressable.Status.Addresses {
+			if addr.Name != nil && *addr.Name == *dest.Ref.Address {
+				return &addr, nil
+			}
+		}
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("address with name %q not found for %+v", *dest.Ref.Address, dest.Ref))
```

Comment:
In one test, we test this scenario, which is great, however, the final error message contains a pointer representation: `... Address:0xc0004c4b`, which isn't very human friendly nor debuggable nor predictable.

I think we could implement the `Stringer` interface (basically a `String() string` method) on the `*KReference` type that will basically gives us a human readable string for a KReference and we can use that String() method every time we need to return an error for a specific `dest.Ref` instead of using `%+v`.

In addition, this would be applicable to errors like:
```
	if addr.URL == nil {
		return nil, apierrs.NewBadRequest(fmt.Sprintf("URL missing in address of %+v", dest.Ref))
	}
	if addr.URL.Host == "" {
		return nil, apierrs.NewBadRequest(fmt.Sprintf("hostname missing in address of %+v", dest.Ref))
	}
```